pkg/ext/stores/tokens/tokens.go

```diff
@@ -0,0 +1,922 @@
+	// `UserPrincipal` is filled in part with standard information, and in part from the
+	// associated `User`s fields.
+
+	attribs, err := t.userAttributeClient.Get(token.Spec.UserID, metav1.GetOptions{})
+	if err != nil {
+		return nil, apierrors.NewInternalError(fmt.Errorf("failed to retrieve user attributes of %s: %w",
+			token.Spec.UserID, err))
+	}
+	if attribs == nil {
+		return nil, apierrors.NewInternalError(fmt.Errorf("failed to get user attributes of %s",
+			token.Spec.UserID))
+	}
+
+	if len(attribs.ExtraByProvider) != 1 {
+		return nil, apierrors.NewInternalError(fmt.Errorf("bad user attributes: bogus ExtraByProvider, empty or ambigous"))
+	}
+
+	user, err := t.userClient.Get(token.Spec.UserID, metav1.GetOptions{})
+	if err != nil {
+		return nil, apierrors.NewInternalError(fmt.Errorf("failed to retrieve user %s: %w",
+			token.Spec.UserID, err))
+	}
+	if user == nil {
+		return nil, apierrors.NewInternalError(fmt.Errorf("failed to get user %s", token.Spec.UserID))
+	}
+
+	// Generate secret and its hash
+
+	tokenValue, hashedValue, err := t.hasher.MakeAndHashSecret()
+	if err != nil {
+		return nil, err
+	}
+
+	// (len == 1) => The single key in the UA map names the auth provider
+	// (and where to look in GPs, if we were using GPs)
+	var authProvider string
+	for ap, _ := range attribs.ExtraByProvider {
+		authProvider = ap
+		break
+	}
+
+	token.Status.TokenHash = hashedValue
+	token.Status.AuthProvider = authProvider
+	token.Status.DisplayName = user.DisplayName
+	token.Status.LoginName = user.Username // See also attribs.ExtraByProvider[ap]["username"][0]
+	token.Status.PrincipalID = attribs.ExtraByProvider[authProvider]["principalid"][0]
+	token.Status.LastUpdateTime = t.timer.Now()
+
+	rest.FillObjectMetaSystemFields(token)
+
+	secret := secretFromToken(token)
+
+	// Save new secret
+	newSecret, err := t.secretClient.Create(secret)
+	if err != nil {
+		if apierrors.IsAlreadyExists(err) {
+			// can happen despite the early check for a pre-existing secret.
+			// something else may have raced us while the secret was assembled.
+			return nil, err
+		}
+		return nil, apierrors.NewInternalError(fmt.Errorf("failed to store token %s: %w",
+			token.Name, err))
+	}
+
+	// Read changes back to return what was truly created, not what we thought we created
+	newToken, err := tokenFromSecret(newSecret)
+	if err != nil {
+		return nil, apierrors.NewInternalError(fmt.Errorf("failed to regenerate token %s: %w",
+			token.Name, err))
+	}
+
+	// users don't care about the hashed value
+	newToken.Status.TokenHash = ""
+	newToken.Status.TokenValue = tokenValue
+	return newToken, nil
+}
+
+func (t *Store) Update(ctx extcore.Context, token *ext.Token, opts *metav1.UpdateOptions) (*ext.Token, error) {
+
+	// Work on the time-to-live value is a bit more complicated. Even the owning user is not
+	// allowed to extend the TTL. Only a `manage-token`-permitted user is allowed to do
+	// that. This is not quite captured by a `hasPermission` flag.
+
```

Comment:
nit: we should avoid (if possible of course) free-floating comments as Go's AST is having a hard time knowing what to do with those.

```suggestion
	// Work on the time-to-live value is a bit more complicated. Even the owning user is not
	// allowed to extend the TTL. Only a `manage-token`-permitted user is allowed to do
	// that. This is not quite captured by a `hasPermission` flag.
```

Reply:
Ir is a comment for me and future me, not for go.
I did not expect go to do anything with it.
Note, I see no diff in the suggested change.